app/src/main/java/com/gdgevents/gdgeventsapp/features/map/presentaion/MapViewModel.kt

```diff
@@ -1,10 +1,139 @@
 package com.gdgevents.gdgeventsapp.features.map.presentaion
 
+import android.location.Address
+import android.location.Geocoder
+import android.util.Log
 import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.gdgevents.gdgeventsapp.core.datastore.domain.UserPreferencesRepo
+import com.google.android.gms.location.FusedLocationProviderClient
+import com.google.android.gms.maps.model.LatLng
 import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.SharedFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
 import javax.inject.Inject
 
+private const val TAG = "MapViewModel"
+
 @HiltViewModel
-class MapViewModel @Inject constructor() : ViewModel() {
-    // add implementation
+class MapViewModel @Inject constructor(
+    private val fusedLocationProviderClient: FusedLocationProviderClient,
+    private val geocoder: Geocoder,
+    private val userPrefRepo: UserPreferencesRepo
+) : ViewModel() {
+    private val _state = MutableStateFlow(MapUiState())
+    val state: StateFlow<MapUiState> = _state
+
+    private val _message = MutableSharedFlow<String>()
+    val message: SharedFlow<String> = _message
+
+    init {
+        getRegionFromLocation(_state.value.marker!!)
```

Comment:
lets not use `_state.value.marker!!` to avoid null pointer exceptions. and use it like this 

```suggestion
_state.value.marker?.let {
getRegionFromLocation(it)
}
        
```